### PR TITLE
Add example for writing SQL analysis using DataFusion structures

### LINKED ISSUE #10871: Add example for writing an SQL analysis pass
### Is your feature request related to a problem or challenge?

We have many good examples in https://github.com/apache/datafusion/tree/3773fb7fb54419f889e7d18b73e9eb48069eb08e/datafusion-examples

However, we don't have anexample  for writing some sort of SQL analysis -- for example  what @LorrensP-2158466  mentions about finding cyclic joins in https://github.com/apache/datafusion/issues/10808#issuecomment-2160928543. 

However, I know of several users of DataFusion doing this (e.g. @sadboy and SDF)

### Describe the solution you'd like

The ideal example I think would be to  add a file to https://github.com/apache/datafusion/tree/3773fb7fb54419f889e7d18b73e9eb48069eb08e/datafusion-examples

`sql_analysis.rs`

Perhaps the example could show how to create LogicalPlans for several sql query texts, and then use DataFusion structures to do an analysis (like maybe join counts, or predicate analysis or something) would be really neat.

I think this would be very valuable

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LorrensP-2158466**
## Which issue does this PR close?

Closes #10871 .

## What changes are included in this PR?

Added an example of an SQL analysis where we count joins in a SQL query.

## Are these changes tested?

Ran it on some other queries and the join counts where right.

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion-examples/examples/sql_analysis.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This example shows how we can use the structures that DataFusion provide to perform
+//! SQL Analysis.
+//! We'll show how we can count the amount of join's in a query as well as how many
```
Comment (alamb) [inline datafusion-examples/examples/sql_analysis.rs:20]: 👍

File: datafusion-examples/examples/sql_analysis.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    // register the tables of the TPC-DS query
+    let tables = tpcds_schemas();
+    for table in tables {
+        ctx.register_table(
+            table.name,
+            Arc::new(MemTable::try_new(Arc::new(table.schema.clone()), vec![])?),
+        )?;
+    }
+
+    let query_88 = "datafusion/core/tests/tpc-ds/88.sql";
```
Comment (alamb) [inline datafusion-examples/examples/sql_analysis.rs:122]: This is very cool -- thank you

This example might actually be easier to understand if we inlined these queries below -- I can do it if you like -- so the example stands on its own

https://github.com/apache/datafusion/blob/main/datafusion/core/tests/tpc-ds/88.sql

File: datafusion-examples/examples/sql_analysis.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+use datafusion::common::Result;
+use datafusion::{
+    datasource::MemTable,
+    execution::context::{SessionConfig, SessionContext},
+};
+use datafusion_common::tree_node::{TreeNode, TreeNodeRecursion};
+use datafusion_expr::LogicalPlan;
+use test_utils::tpcds::tpcds_schemas;
+
+/// Counts the total amount of joins in a plan
```
Comment (alamb) [inline datafusion-examples/examples/sql_analysis.rs:34]:
```suggestion
/// Counts the total number of joins in a plan
```

File: datafusion-examples/examples/sql_analysis.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        Ok(TreeNodeRecursion::Continue)
+    })
+    .unwrap();
+
+    total
+}
+
+/// Counts the total amount of joins and collects every join tree with it's join count
+/// The list of groups sums to the total count
+fn count_trees(plan: &LogicalPlan) -> (usize, Vec<usize>) {
```
Comment (alamb) [inline datafusion-examples/examples/sql_analysis.rs:53]: I found it a little hard to follow what qualified as a join tree and when they are different

maybe a definition and few text examples would help, like

```
A join tree is defined as the largest sub trees which consistt entirely of joins

For example

  JOIN
  /  \
A   JOIN
     /  \
    B    C

has a single join tree (A-B-C) but this plan:

  JOIN
  /  \
A   GROUP
       |
      JOIN
     /  \
    B    C

Has two join trees (A, B-C)
```

**Comment by LorrensP-2158466** [comment]
It's currently midnight where I live, so I can do it tomorrow. If you have free time you're free to make these changes.
Thanks for the reply and help!

**Comment by alamb** [comment]
> It's currently midnight where I live, so I can do it tomorrow. If you have free time you're free to make these changes. Thanks for the reply and help!

No worries! In general we operate asynchronously, so there is no need to an immediate response. Thank you for making the changes -- I am taking a second look now

**Review by alamb — APPROVED** [review]
This is amazing -- thank you so much @LorrensP-2158466 

It is really neat to see how to use the TreeNode API from @peter-toth  and others to write non trivial analyses that walk the tree. 

I took the liberty of merging this branch up to main and fixing the Cargo.toml formatting, as well doing some small textual cleanups

I plan to merge it once the CI checks complete. 

Again, thank you, this is so cool

**Comment by alamb** [comment]
Thanks again @LorrensP-2158466

**Comment by LorrensP-2158466** [comment]
Glad I could help, it was a lot fun!
